src/lib/network_pool/transacation_pool_reorg_tests.ml

```diff
@@ -0,0 +1,433 @@
+open Core
+open Async
+open Mina_base
+open Mina_transaction
+open Inline_test_quiet_logs
+open Signature_lib
+open Test_utils
+
+let%test_module "transaction pool reorg tests" =
+( module struct
+  
+
+module Account_spec = struct
+  type t = {
+    key_idx : int
+    ; balance: int
+    ; nonce: int
+  } [@@deriving sexp]
+
+  let rec find_in_test_keys ?(n=0) (x:Keypair.t) =
+    if Keypair.equal test_keys.(n) x then n 
+    else find_in_test_keys x ~n:(n+1)
+
+  let create key_idx balance nonce = 
+    {
+      key_idx
+      ; balance 
+      ; nonce
+    }
+
+  let of_ledger_row ledger_row = 
+    let (kp,balance,nonce,_)  = ledger_row in
+    create 
+       (find_in_test_keys kp) 
+       (Currency.Amount.to_nanomina_int balance)
+       (Unsigned.UInt32.to_int nonce)
+  
+  let apply_payment amount fee t = 
+    create t.key_idx (t.balance - amount - fee) (t.nonce + 1)
+
+  let apply_zkapp fee t = create t.key_idx (t.balance - fee) (t.nonce + 1)
+
+  let to_key_and_nonce t = 
+    (Public_key.compress test_keys.(t.key_idx).public_key, t.nonce)
+
+end
+
+module Command_spec = struct
+
+  type t = 
+    | Payment of { sender : Account_spec.t
+      ; receiver_idx: int
+      ; fee: int
+      ; amount: int
+    } 
+    | Zkapp_blocking_send of { sender : Account_spec.t
+      ; fee: int
+    }
+  [@@deriving sexp]
+
+  let gen_zkapp_blocking_send (spec:Account_spec.t array)
+   = 
+    let open Quickcheck.Generator.Let_syntax in
+    let%bind (random_idx, account_spec) = Array.mapi ~f:(fun i e -> (i,e)) spec |> Quickcheck_lib.of_array in
+    let new_account_spec = Account_spec.apply_zkapp minimum_fee account_spec in 
+    Array.set spec random_idx new_account_spec;
+    return (Zkapp_blocking_send {
+      sender = account_spec
+      ; fee = minimum_fee
+    }) 
+
+  let gen_single_from (spec:Account_spec.t array) (idx,account_spec) = 
+      let open Quickcheck.Generator.Let_syntax in
+        let%bind receiver_idx = test_keys |> Array.mapi ~f:(fun i _-> i) |> Quickcheck_lib.of_array in
+        let%bind amount = Int.gen_incl 5_000_000_000_000 10_000_000_000_000  in
+        let new_account_spec = Account_spec.apply_payment amount minimum_fee account_spec in 
+        Array.set spec idx new_account_spec;
```

Comment:
What if balance goes negative? Perhaps we set amount to `0` if amount is less than balance. And if balance is less than min fee, throw an exception.

Reply:
Case of fee applies to zkapp tx above as well